pallene/to_ir.lua

```diff
@@ -209,66 +209,150 @@ function ToIR:convert_stat(cmds, stat)
         local decls = stat.decls
         local exps = stat.exps
 
-        -- for-in loops are desugared into regurlar loops before compiling.
-        -- For example, a loop like this:
-        --- ```
-        -- for a: T1, b: T2 in <RHS> do
-        --     <loop body>
-        -- end
-        ---```
-        -- is compiled as if the following was written instead:
-        --```
-        -- local iter, st, a_any, b_any
-        -- iter, st, ctrl = RHS[1], RHS[2], RHS[3]
-        -- while true do
-        --   a_any, b_any = iter(st, ctrl)
-        --   if a_any == nil then break end
-        --   local a = a_any as T1
-        --   local b = b_any as T2
-        --   <loop body>
-        -- end
-        -- ```
-
-        local v_iter = ir.add_local(self.func, "$iter", exps[1]._type)
-        self:exp_to_assignment(cmds, v_iter, exps[1])
-        local v_state = ir.add_local(self.func, "$st", exps[2]._type)
-        self:exp_to_assignment(cmds, v_state, exps[2])
-
-        local v_lhs_dyn = {}
-        for _, decl in ipairs(decls) do
-            local v = ir.add_local(self.func, "$" .. decl.name .. "_dyn", types.T.Any())
-            table.insert(v_lhs_dyn, v)
-        end
+        local e1 = exps[1]
+        local is_ipairs = (
+            e1._tag == "ast.Exp.CallFunc" and
+            e1.exp._tag == "ast.Exp.Var" and
+            e1.exp.var._name._tag == "checker.Name.Builtin" and
+            e1.exp.var._name.name == "ipairs")
+
+
+        if is_ipairs then
+            -- `ipairs` are desugared down to regular for-loops
+            -- ```
+            -- for i: T1, x: T2 in ipairs(xs) do
+            --   <loop body>
+            -- end
+            -- ```
+            -- would get compiled down to:
+            -- ```
+            -- local i_num: integer = 1
+            -- while true do
+            --   local x_dyn = xs[i_num]
+            --   if x_dyn == nil then
+            --     break
+            --   end
+            --   local i = i_num as T1
+            --   local x = x_dyn as T2 
+            --   <loop body>
+            --   i_num = i_num + 1
+            -- end
+            -- ```
+
+            -- the table passed as argument to `ipairs`
+            local arr =  exps[2].call_exp.args[1]
```

Comment:
True, the error for ipairs won't make a lot of sense, I'll open an issue.

Reply:
> ipairs is called with one argument, which is a function call returning one value

Should this really be an error if the value is a table value?

Reply:
Now that you mention it, I agree that it would be redundant to test what happens the argument to ipairs is a function call.
The important thing to test is what happens if we call ipairs with multiple arguments, where the additional arguments might have side effects.

    ipairs(f(), g())
    
In the current implementation, I believe the call to `g()` would not be generated at all, which can be a problem if the function call does something (like printing some output, or modifying a mutable table).

Reply:
Just to be clear, I think that what should be done in that example with the `g()` is raise a compile-time type error saying that it expected 1 argument but found 2. Then, in the `to_ir` we can put an assert saying that `#args == 1`.

The example was just to motivate why I think it is important to assert that `#args == 1` in the `to_ir`. Because if it there were more arguments, the current implementation would be buggy. Since this is an important assumption, might as well assert it to be sure that it holds.